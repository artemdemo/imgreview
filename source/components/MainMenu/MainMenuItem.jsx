import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import './MainMenuItem.less';

class MainMenuItem extends React.PureComponent {
    render() {
        const { disabled, right, onClick } = this.props;
        const buttonClass = classnames({
            'main-menu-item': true,
            'main-menu-item_disabled': disabled,
            'main-menu-item_right': right,
        });
        return (
            <button
                className={buttonClass}
                disabled={disabled}
                type='button'
                onClick={onClick}
            >
                {this.props.children}
            </button>
        );
    }
}

MainMenuItem.propTypes = {
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
    right: PropTypes.bool,
};

MainMenuItem.defaultProps = {
    onClick: null,
    disabled: null,
    right: false,
};

export default MainMenuItem;
