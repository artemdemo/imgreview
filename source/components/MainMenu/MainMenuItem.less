@import '../../styles/functions';
@import '../../styles/variables';

@horizontal-padding: 4px;
@menu-item-border-color: #b8b8b8;

.main-menu-item {
    .resetButton();
    background-color: @main-menu-color;
    padding: @horizontal-padding 6px;
    border: 1px solid @menu-item-border-color;
    border-radius: 3px;
    float: left;
    margin-right: 5px;
    outline: none;
    display: flex;
    position: relative;
}
    .main-menu-item_disabled {
        opacity: 0.4;
        cursor: not-allowed;
    }

    .main-menu-item__content {
        flex-grow: 1;
        min-width: 30px - 2 * @horizontal-padding;
    }

    .main-menu-item__caret {
        flex-grow: 0;
        padding-left: 5px;
    }

.main-menu-item-submenu {
    position: absolute;
    top: 100%;
    display: none;
    padding-top: 5px;
    z-index: @main-menu-z-index;

    .main-menu-item:hover & {
        display: block;
    }
}
    .main-menu-item-submenu__item {
        padding: 4px 10px;
        border-bottom: 1px dashed @menu-item-border-color;

        &:last-child {
            border-bottom: none;
        }
    }

.main-menu-item-submenu-body {
    background-color: @main-menu-color;
    border: 1px solid @menu-item-border-color;
    border-radius: 3px;
    padding: 3px;
}
