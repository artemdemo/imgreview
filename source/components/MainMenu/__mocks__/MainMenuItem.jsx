import React from 'react';

const MainMenuItem = props => (
    <div data-mock='Popup'>
        {props.children}
    </div>
);

export default MainMenuItem;
