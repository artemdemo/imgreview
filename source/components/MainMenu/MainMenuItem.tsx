import React from 'react';
import classnames from 'classnames';
import Icon from '../Icon/Icon';
import SubMenu, { TSubmenuData } from './SubMenu';

import './MainMenuItem.less';

type Props = {
    subMenu: TSubmenuData;
    disabled: boolean;
    onClick: (e?: any) => void;
};

class MainMenuItem extends React.PureComponent<Props> {
    static readonly defaultProps = {
        onClick: null,
        disabled: null,
        subMenu: [],
    };

    renderCaret() {
        const { subMenu } = this.props;
        if (subMenu.length > 0) {
            return (
                <span className='main-menu-item__caret'>
                    <Icon name='caret-down' />
                </span>
            );
        }

        return null;
    }

    renderSubMenu() {
        const { subMenu } = this.props;
        if (subMenu.length > 0) {
            return (
                <div className='main-menu-item-submenu'>
                    <SubMenu data={subMenu} />
                </div>
            );
        }

        return null;
    }

    render() {
        const { disabled, onClick } = this.props;
        const buttonClass = classnames({
            'main-menu-item': true,
            'main-menu-item_disabled': disabled,
        });
        return (
            <button
                className={buttonClass}
                disabled={disabled}
                type='button'
                onClick={onClick}
            >
                <span
                    className='main-menu-item__content'
                >
                    {this.props.children}
                </span>
                {this.renderCaret()}
                {this.renderSubMenu()}
            </button>
        );
    }
}

export default MainMenuItem;
