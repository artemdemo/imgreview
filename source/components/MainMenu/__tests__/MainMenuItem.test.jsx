import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import MainMenuItem from '../MainMenuItem';

jest.mock('../../Icon/Icon');
jest.mock('../SubMenu');

describe('MainMenuItem', () => {
    it('simple render', () => {
        const tree = renderer.create(
            <MainMenuItem />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should be disabled', () => {
        const tree = renderer.create(
            <MainMenuItem disabled />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should handle onClick', () => {
        const onClickMock = jest.fn();
        const wrapper = mount(
            <MainMenuItem onClick={onClickMock}>
                Button text
            </MainMenuItem>
        );
        wrapper.find('button').simulate('click');
        expect(onClickMock).toBeCalled();
    });

    it('should handle subMenu', () => {
        const wrapper = mount(
            <MainMenuItem
                subMenu={[
                    {text: 'some text - 1', value: 1},
                    {text: 'some text - 2', value: 2},
                ]}
            >
                Button text
            </MainMenuItem>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
