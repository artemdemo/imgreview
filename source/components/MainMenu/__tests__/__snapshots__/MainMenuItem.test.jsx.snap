// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainMenuItem should be disabled 1`] = `
<button
  className="main-menu-item main-menu-item_disabled"
  disabled={true}
  onClick={[Function]}
  type="button"
/>
`;

exports[`MainMenuItem simple render 1`] = `
<button
  className="main-menu-item"
  disabled={null}
  onClick={[Function]}
  type="button"
/>
`;
